py/qstr.c

```diff
@@ -179,13 +228,45 @@ STATIC qstr qstr_add(mp_uint_t hash, mp_uint_t len, const char *q_ptr) {
     return MP_STATE_VM(last_pool)->total_prev_len + at;
 }
 
+#define MP_QSTR_SEARCH_THRESHOLD 10
+
 qstr qstr_find_strn(const char *str, size_t str_len) {
-    // work out hash of str
     mp_uint_t str_hash = qstr_compute_hash((const byte *)str, str_len);
 
     // search pools for the data
     for (const qstr_pool_t *pool = MP_STATE_VM(last_pool); pool != NULL; pool = pool->prev) {
-        for (mp_uint_t at = 0, top = pool->len; at < top; at++) {
+        size_t low = 0;
+        size_t high = pool->len - 1;
+
+        // binary search inside the pool
+        if (pool->sorted) {
+            while (high - low > MP_QSTR_SEARCH_THRESHOLD) {
+                size_t mid = (low + high + 1) / 2;
+                int cmp = pool->hashes[mid] - str_hash;
+                if (cmp == 0) cmp = pool->lengths[mid] - str_len;
+                if (cmp < 0) {
+                    low = mid;
+                } else if (cmp > 0) {
+                    high = mid;
+                } else {
+                    // avoid a rare (hash,len) collisions
+                    while (MP_UNLIKELY(
+                        pool->lengths[mid] != str_len ||
```

Comment:
I don't think you need this check on length.  If the code gets here then `cmp == 0`, which can only happen if the lengths are equal (and also the hashes are equal).

Reply:
And wouldn't it be simpler (smaller code, probably no effect on performance) to just break here and drop down into the sequential search?

Reply:
You are right - this can be simplified.  
Done in 3bc257c4b24d5d2e4c58c13a91a31f1a58306bdd.